monai/networks/trt_wrapper.py

```diff
@@ -0,0 +1,440 @@
+    from polygraphy.backend.trt import (
+        CreateConfig,
+        Profile,
+        engine_from_bytes,
+        engine_from_network,
+        network_from_onnx_path,
+        save_engine,
+    )
+
+trt, trt_imported = optional_import("tensorrt")
+cudart, _ = optional_import("cuda.cudart")
+
+LOGGER = get_logger("run_cmd")
+
+lock_sm = threading.Lock()
+
+
+# Map of TRT dtype -> Torch dtype
+def trt_to_torch_dtype_dict():
+    return {
+        trt.int32: torch.int32,
+        trt.float32: torch.float32,
+        trt.float16: torch.float16,
+        trt.bfloat16: torch.float16,
+        trt.int64: torch.int64,
+        trt.int8: torch.int8,
+        trt.bool: torch.bool,
+    }
+
+
+def get_dynamic_axes(profiles):
+    """
+    Given [[min,opt,max],...] list of profile dimensions,
+    this method calculates dynamic_axes to use in onnx.export()
+    """
+    dynamic_axes: dict[str, list[int]] = {}
+    if not profiles:
+        return dynamic_axes
+    for profile in profiles:
+        for key in profile:
+            axes = []
+            vals = profile[key]
+            for i in range(len(vals[0])):
+                if vals[0][i] != vals[2][i]:
+                    axes.append(i)
+            if len(axes) > 0:
+                dynamic_axes[key] = axes
+    return dynamic_axes
+
+
+def cuassert(cuda_ret):
+    """
+    Error reporting method for CUDA calls
+    """
+    err = cuda_ret[0]
+    if err != 0:
+        raise RuntimeError(f"CUDA ERROR: {err}")
+    if len(cuda_ret) > 1:
+        return cuda_ret[1]
+    return None
+
+
+class ShapeError(Exception):
+    """
+    Exception class to report errors from setting TRT plan input shapes
+    """
+
+    pass
+
+
+class TRTEngine:
+    """
+    An auxiliary class to implement running of TRT optimized engines
+
+    """
+
+    def __init__(self, engine_path):
+        """
+        Loads serialized engine, creates execution context and activates it
+        """
+        self.engine_path = engine_path
+        LOGGER.info(f"Loading TensorRT engine: {self.engine_path}")
+        self.engine = engine_from_bytes(bytes_from_path(self.engine_path))
+        self.tensors = OrderedDict()
+        self.cuda_graph_instance = None  # cuda graph
+        self.context = self.engine.create_execution_context()
+        self.input_names = []
+        self.output_names = []
+        self.dtypes = []
+        self.cur_profile = 0
+        dtype_dict = trt_to_torch_dtype_dict()
+        for idx in range(self.engine.num_io_tensors):
+            binding = self.engine[idx]
+            if self.engine.get_tensor_mode(binding) == trt.TensorIOMode.INPUT:
+                self.input_names.append(binding)
+            elif self.engine.get_tensor_mode(binding) == trt.TensorIOMode.OUTPUT:
+                self.output_names.append(binding)
+                dtype = dtype_dict[self.engine.get_tensor_dtype(binding)]
+                self.dtypes.append(dtype)
+
+    def allocate_buffers(self, device):
```

Comment:
Hi @borisfom , thank you for the PR! 
I am a beginner in trt and I may be totally wrong here: this function seems not check if the buffers are already allocated before allocating new ones. Could this lead to memory leaks if the `forward` method is called multiple times?

Reply:
Well this function reassigns output buffers (self.tensors) so it does handle release of previously allocated buffers automatically. Thanks for pointing this out though - it can be optimized out for the case of constant size output when the size is not changing (ever) - the gain seems to be really small for those nets but I am adding it anyway.

Reply:
Thanks for the checking!